File: leetcode/Sort/leetcode-147/147.go
package sort

// ListNode : Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func insertionSortList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	ret := &ListNode{0, nil}
	cur := head
	for cur != nil {
		tmp := cur.Next
		insertList(cur, ret)
		cur = tmp
	}
	return ret.Next
}

func insertList(node, pre *ListNode) {
	head := pre.Next
	for head != nil && node.Val > head.Val {
		pre = head
		head = head.Next
	}
	pre.Next = node
	node.Next = head
}

File: leetcode/Sort/leetcode-147/147.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode insertionSortList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode ret = new ListNode(0);

        ListNode cur = head;

        while (cur != null) {
            ListNode tmp = cur.next;
            insertList(cur, ret);
            cur = tmp;
        }

        return ret.next;
    }

    public void insertList(ListNode node, ListNode pre) {
        ListNode head = pre.next;
        while (head != null && node.val > head.val) {
            pre = head;
            head = head.next;
        }
        pre.next = node;
        node.next = head;
    }
}